File: webapp/lib/decision.rb
# frozen_string_literal: true

# A choice the game is waiting for a player to make. Holds no logic — it only
# describes what is being decided, by whom, and the available options.
#
# kind:
#   :choose_boss       pick one of two drawn boss cards (the other is discarded)
#   :place_first_room  pick a room from hand to place beside the boss (Setup)
#   :discard_room      pick a room from hand to throw away (Build)
#   :build_room        pick a room from hand to place, or skip (Build)
class Decision
  KINDS = %i[choose_boss place_first_room discard_room build_room].freeze

  attr_reader :kind, :player, :options, :allow_skip

  # @param options [Array] the cards the player may choose among
  # @param allow_skip [Boolean] whether choosing nothing is permitted
  def initialize(kind:, player:, options:, allow_skip: false)
    raise ArgumentError, "unknown decision kind: #{kind}" unless KINDS.include?(kind)

    @kind = kind
    @player = player
    @options = (options || []).freeze
    @allow_skip = allow_skip
    freeze
  end

  def prompt
    case kind
    when :choose_boss
      "#{player.name}: choose your boss (the other is discarded)"
    when :place_first_room
      "#{player.name}: choose a room to place beside your boss"
    when :discard_room
      "#{player.name}: discard a room card"
    when :build_room
      "#{player.name}: choose a room to add to your dungeon, or build nothing"
    end
  end
end

File: android/app/src/main/java/com/dungeonboss/game/Decision.kt
package com.dungeonboss.game

import com.dungeonboss.model.Card

/** The kinds of choice the game can wait on a player to make. */
enum class DecisionKind { CHOOSE_BOSS, PLACE_FIRST_ROOM, DISCARD_ROOMS, BUILD_ROOM }

/**
 * A choice the game is waiting for a player to make. Holds no logic — it only
 * describes what is being decided, by whom, and the available options.
 *
 *   CHOOSE_BOSS       pick one of two drawn boss cards (the other is discarded)
 *   PLACE_FIRST_ROOM  pick a room from hand to place into any of the 5 slots (Setup)
 *   DISCARD_ROOMS     pick 0–2 room cards from hand to throw away (Discard phase)
 *   BUILD_ROOM        pick a card from hand to play (into a slot, or to upgrade a
 *                     room), or skip (Build)
 */
class Decision(
    val kind: DecisionKind,
    val player: Player,
    options: List<Card>,
    val allowSkip: Boolean = false
) {
    val options: List<Card> = options.toList()

    fun prompt(): String = when (kind) {
        DecisionKind.CHOOSE_BOSS ->
            "${player.name}: choose your boss (the other is discarded)"
        DecisionKind.PLACE_FIRST_ROOM ->
            "${player.name}: choose a room and a slot to place it in"
        DecisionKind.DISCARD_ROOMS ->
            "${player.name}: discard 0–2 room cards (optional)"
        DecisionKind.BUILD_ROOM ->
            "${player.name}: place or upgrade a room, or build nothing"
    }
}
